### PR TITLE
Remove Carbon from Mac Auto-Type

### LINKED ISSUE #3310: Replace Carbon HotKey registration with NSEvent API
[TIP]:  # ( Provide a general summary of the issue in the title above ^^ )
[TIP]:  # ( DO NOT include screenshots of your actual database! )

## Expected Behavior
[NOTE]: # ( Tell us what you expected to happen )
Use modern API's on macOS

## Current Behavior
[NOTE]: # ( Tell us what actually happens )
We use Carbon to register the global hotkey. This method is deprecated since 10.8.

## Possible Solution
[NOTE]: # ( Not required, but suggest a fix/reason for the bug )
https://developer.apple.com/documentation/appkit/nsevent/1535472-addglobalmonitorforeventsmatchin?language=objc

Replace `static uint16 qtToNativeKeyCode(Qt::Key key);` with `static CGKeyCode qtToNativeKeyCode(Qt::Key key);`

Remove `-framework Carbon` from mac-autotype CMakeLists

### PR DESCRIPTION
**Pull request by droidmonkey**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Refactor (significant modification to existing code)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
* Fix #3310
* Replaces Carbon-based API for global hotkey handling with NSEvent
* Thanks to @rockihack for his code submission

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested on Mojave

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: src/autotype/mac/AutoTypeMac.cpp [hunk 1]
```diff
@@ -17,31 +17,23 @@
...
 {
-    EventTypeSpec eventSpec;
-    eventSpec.eventClass = kEventClassKeyboard;
-    eventSpec.eventKind = kEventHotKeyPressed;
-
-    ::InstallApplicationEventHandler(AutoTypePlatformMac::hotkeyHandler, 1, &eventSpec, this, nullptr);
 }
 
 //
 // Keepassx requires mac os 10.7
```
Comment (phoerious) [inline src/autotype/mac/AutoTypeMac.cpp]: Update this comment?

File: src/autotype/mac/AutoTypeMac.cpp [hunk 2]
```diff
@@ -116,29 +110,34 @@ QString AutoTypePlatformMac::activeWindowTitle()
...
-    }
-
+    CGEventFlags nativeModifiers = qtToNativeModifiers(modifiers);
+    m_globalMonitor =
+        macUtils()->addGlobalMonitor(nativeKeyCode, nativeModifiers, this, AutoTypePlatformMac::hotkeyHandler);
     return true;
 }
 
+//
+// Hotkey Handler
```
Comment (phoerious) [inline src/autotype/mac/AutoTypeMac.cpp]: I generally don't like this comment style. Use proper Doc blocks.

File: src/autotype/mac/AutoTypeMac.cpp [hunk 3]
```diff
@@ -116,29 +110,34 @@ QString AutoTypePlatformMac::activeWindowTitle()
...
+//
+void AutoTypePlatformMac::hotkeyHandler(void* userData)
+{
+    AutoTypePlatformMac* self = static_cast<AutoTypePlatformMac*>(userData);
```
Comment (phoerious) [inline src/autotype/mac/AutoTypeMac.cpp]: auto if type is on the right-hand side.

File: src/autotype/mac/AutoTypeMac.cpp [hunk 4]
```diff
@@ -217,214 +216,201 @@ void AutoTypePlatformMac::sendKey(Qt::Key key, bool isKeyDown, Qt::KeyboardModif
...
+        return kVK_F13;
+    case Qt::Key_F14:
+        return kVK_F14;
+    case Qt::Key_F15:
+        return kVK_F15;
+    case Qt::Key_F16:
+        return kVK_F16;
+
+    default:
+        Q_ASSERT(false);
```
Comment (phoerious) [inline src/autotype/mac/AutoTypeMac.cpp]: Does this actually justify an assertion?

File: src/autotype/mac/AutoTypeMac.cpp [hunk 5]
```diff
@@ -438,8 +424,7 @@ int AutoTypePlatformMac::windowLayer(CFDictionaryRef window)
     int layer;
 
     CFNumberRef layerRef = static_cast<CFNumberRef>(::CFDictionaryGetValue(window, kCGWindowLayer));
-    if (layerRef != nullptr
-            && ::CFNumberGetValue(layerRef, kCFNumberIntType, &layer)) {
+    if (layerRef != nullptr && ::CFNumberGetValue(layerRef, kCFNumberIntType, &layer)) {
```
Comment (phoerious) [inline src/autotype/mac/AutoTypeMac.cpp]: if (layerRef && ::CFNumberGetValue...)

File: src/autotype/mac/AutoTypeMac.cpp [hunk 6]
```diff
@@ -455,32 +440,13 @@ QString AutoTypePlatformMac::windowTitle(CFDictionaryRef window)
     QString title;
 
     CFStringRef titleRef = static_cast<CFStringRef>(::CFDictionaryGetValue(window, kCGWindowName));
-    if (titleRef != nullptr
-            && ::CFStringGetCString(titleRef, buffer, MAX_WINDOW_TITLE_LENGTH, kCFStringEncodingUTF8)) {
+    if (titleRef != nullptr && ::CFStringGetCString(titleRef, buffer, MAX_WINDOW_TITLE_LENGTH, kCFStringEncodingUTF8)) {
```
Comment (phoerious) [inline src/autotype/mac/AutoTypeMac.cpp]: Ditto

File: src/gui/macutils/AppKitImpl.mm [hunk 7]
```diff
@@ -118,6 +118,31 @@ - (void) userSwitchHandler:(NSNotification*) notification
     }
 }
 
+//
+// Add global event monitor
+//
+- (id) addGlobalMonitor:(NSEventMask) mask handler:(void (^)(NSEvent *)) handler
```
Comment (phoerious) [inline src/gui/macutils/AppKitImpl.mm]: NSEvent*

File: src/gui/macutils/AppKitImpl.mm [hunk 8]
```diff
@@ -118,6 +118,31 @@ - (void) userSwitchHandler:(NSNotification*) notification
...
+{
+    [NSEvent removeMonitor:monitor];
+}
+
+//
+// Check if accessibility is enabled, may show an popup asking for permissions
+//
+- (bool) enableAccessibility
+{
+    NSDictionary *opts = @{static_cast<id>(kAXTrustedCheckOptionPrompt): @YES};
```
Comment (phoerious) [inline src/gui/macutils/AppKitImpl.mm]: NSDictionary* opts

There are more things like that in this file (and the next).

**Comment by droidmonkey** [comment]
@phoerious nice review, you got me!

File: src/autotype/mac/AutoTypeMac.cpp [hunk 9]
```diff
@@ -116,29 +110,34 @@ QString AutoTypePlatformMac::activeWindowTitle()
...
 //
 // Unregister global hotkey
 //
 void AutoTypePlatformMac::unregisterGlobalShortcut(Qt::Key key, Qt::KeyboardModifiers modifiers)
 {
     Q_UNUSED(key);
     Q_UNUSED(modifiers);
-
-    ::UnregisterEventHotKey(m_hotkeyRef);
+    macUtils()->removeGlobalMonitor(m_globalMonitor);
```
Comment (rockihack) [inline src/autotype/mac/AutoTypeMac.cpp]: I noticed a problem with my implementation here. A segmentation fault occurs in `[NSEvent removeMonitor:]` if m_globalMonitor is non null and is passed more than once to removeMonitor.

Fix:
```
if (m_globalMonitor) {
    macUtils()->removeGlobalMonitor(m_globalMonitor);
    m_globalMonitor = nullptr;
}
```

Comment (droidmonkey) [reply src/autotype/mac/AutoTypeMac.cpp]: Yes I have just done this haha, found that same error

**Comment by droidmonkey** [comment]
changes made

**Comment by georgesnow** [comment]
I get the impression based on the lack of response to this issue: #3393

The  NSEvent replacement is staying and not reverting back to the Carbon API (which is **not** deprecated) 
As I stated on the other issue KeePassXC is viewing all key presses when using the NSEvent API, which would appropriate for an app like TextExpander, but not for hotkeys. 

https://developer.apple.com/documentation/appkit/nsevent?language=objc

> "The **NSEvent object contains** pertinent information about each event, such as where the cursor was located or **which character was typed**...global event monitor that **allows you to monitor events in other applications,**"

**Comment by droidmonkey** [comment]
I understand your frustrations but we made a change for a reason. No additional security requirements are necessary to support NSEvent versus Carbon. I can consider a reversion of this PR, but it won't happen soon.

**Comment by georgesnow** [comment]
@droidmonkey understood. 

Accessibility is needed for Autotype to work in general, but the inherent difference is KPXC is registering to view all keystrokes under that security permission. The Carbon API doesn't require the to register to **view all keystrokes**. It is registering the hotkey with the OS. My point here is if the app doesn't need that level of access and there is alternative _it seems best to use the least amount of access._ 

And comes down to two parts:
1-Security permissions <=this is the same regardless for NSEvent and Carbon to use GAutotype/Autotype
2-How the hotkey event is being handled <=**this is a coding difference as stated above**


https://www.macworld.com/article/3311982/the-difference-between-accessibility-and-full-disk-access.html

> Apple requires explicit permission, because it’s just these kinds of features that can be leveraged and abused by simpler malware that doesn’t dig deeply into exploiting the system, but could, for instance, try to capture your keystrokes.

**Comment by parodymshifter** [comment]
On Mojave 10.14.6 the change to autotype hotkey handling is dropping the 1st character of passwords in Terminal windows.  It seems to work correctly in a text editor window. The breakage of the function in Terminal windows makes the hotkey feature nearly useless for me.

**Comment by georgesnow** [comment]
This is Most likely because the Nsevent api doesn’t allow  (and isn’t) interrupting or blocking the event from passing. it only observes it. Carbon API stop it’s from posting to other applications

**Comment by droidmonkey** [comment]
Please open an issue about this

**Pull request merged**